packages/@aws-cdk/core/lib/private/tree-metadata.ts

```diff
@@ -20,37 +21,42 @@ export class TreeMetadata extends Construct {
   }
 
   protected synthesize(session: ISynthesisSession) {
-    const lookup: { [path: string]: Node } = { };
+    try {
+      const lookup: { [path: string]: Node } = { };
 
-    const visit = (construct: IConstruct): Node => {
-      const children = construct.node.children.map(visit);
-      const childrenMap = children.reduce((map, child) => Object.assign(map, { [child.id]: child }), {});
-      const node: Node = {
-        id: construct.node.id || 'App',
-        path: construct.node.path,
-        children: children.length === 0 ? undefined : childrenMap,
-        attributes: this.getAttributes(construct)
-      };
+      const visit = (construct: IConstruct): Node => {
+        const children = construct.node.children.map(visit);
+        const childrenMap = children.reduce((map, child) => Object.assign(map, { [child.id]: child }), {});
+        const node: Node = {
+          id: construct.node.id || 'App',
+          path: construct.node.path,
+          children: children.length === 0 ? undefined : childrenMap,
+          attributes: this.getAttributes(construct)
+        };
 
-      lookup[node.path] = node;
+        lookup[node.path] = node;
 
-      return node;
-    };
+        return node;
+      };
 
-    const tree = {
-      version: 'tree-0.1',
-      tree: visit(this.node.root),
-    };
+      const tree = {
+        version: 'tree-0.1',
+        tree: visit(this.node.root),
+      };
 
-    const builder = session.assembly;
-    fs.writeFileSync(path.join(builder.outdir, FILE_PATH), JSON.stringify(tree, undefined, 2), { encoding: 'utf-8' });
+      const builder = session.assembly;
+      fs.writeFileSync(path.join(builder.outdir, FILE_PATH), JSON.stringify(tree, undefined, 2), { encoding: 'utf-8' });
 
-    builder.addArtifact('Tree', {
-      type: ArtifactType.CDK_TREE,
-      properties: {
-        file: FILE_PATH
-      }
-    });
+      builder.addArtifact('Tree', {
+        type: ArtifactType.CDK_TREE,
+        properties: {
+          file: FILE_PATH
+        }
+      });
+    } catch (e) {
+      // Prevent errors here to fail synthesis, until this module stays @experimental.
+      this.node.addWarning(`tree.json could not be generated - ${e}`);
+    }
   }
 
   private getAttributes(construct: IConstruct): { [key: string]: any } | undefined {
```

Comment:
Rename this method to `synthesizeAttributes` to reflect that it is executed during synthesis (and therefore a "resolve" is legitimate).